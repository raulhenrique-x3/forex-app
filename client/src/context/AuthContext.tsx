import React, { createContext, useMemo, useState } from "react";

interface IProps {
  isAuthenticated: boolean;
  login: (value: boolean) => void;
  logout: () => void;
}

export const AuthContext = createContext<IProps>({
  isAuthenticated: false,
  login() {},
  logout() {},
});

interface IChildren {
  children: React.ReactNode;
}

export function AuthContextProvider({ children }: IChildren) {
  const [isAuthenticated, setIsAuthenticated] = useState<boolean>(false);

  const login = (value: boolean) => {
    setIsAuthenticated(value);
  };

  const logout = () => {
    setIsAuthenticated(false);
  };

  const value = useMemo(() => ({ isAuthenticated, login, logout }), [isAuthenticated]);
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
}

export const AuthConsumer = AuthContext.Consumer;
