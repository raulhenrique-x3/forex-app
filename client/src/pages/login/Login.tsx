import {
  Box,
  Button,
  Card,
  CardBody,
  CardHeader,
  Container,
  FormLabel,
  Input,
  Stack,
  StackDivider,
  Text,
  useToast,
} from "@chakra-ui/react";
import useAuth from "../../hooks/useAuth";
import { useEffect, useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import io from "socket.io-client";
import axios from "axios";

export const Login = () => {
  const socket = io("http://localhost:5000", { autoConnect: false });
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const toast = useToast();
  const { login } = useAuth();
  const navigate = useNavigate();

  useEffect(() => {
    socket.on("success_login", () => {
      login(true);
    });

    socket.on("disconnect", () => {
      login(false);
    });

    return () => {
      socket.off("connect");
      socket.off("disconnect");
    };
  }, []);

  const userLogin = () => {
    axios
      .post("http://localhost:5000/session/", { userEmail: email, userPassword: password })
      .then((res) => {
        socket.emit("user_login");
        login(true);
        toast({
          title: "Logged",
          status: "success",
          isClosable: true,
        });
        navigate(`/home/${res.data._id}`);
      })
      .catch((err) => {
        toast({
          title: err.response.data.message,
          status: "error",
          isClosable: true,
        });
      });
  };
  return (
    <Box
      backgroundColor={"#cccccc"}
      width={"100vw"}
      height={"100vh"}
      display="flex"
      alignItems="center"
      justifyContent="center"
      flexDirection="column"
    >
      <Container>
        <Card>
          <CardHeader>
            <Text fontSize={"1.5rem"} fontWeight={"bold"}>
              Login
            </Text>
          </CardHeader>
          <CardBody>
            <Stack divider={<StackDivider />} spacing="4">
              <Box>
                <FormLabel>Email</FormLabel>
                <Input type="text" placeholder="Enter your email" onChange={(e) => setEmail(e.target.value)} />
              </Box>
              <Box>
                <Text>Password</Text>
                <Input
                  type="password"
                  id="password"
                  placeholder="Enter your password"
                  onChange={(e) => setPassword(e.target.value)}
                />
              </Box>
              <Box display={"flex"} flexDirection={"column"}>
                <Button color={"#ffffff"} mt={4} backgroundColor={"#8236FD"} onClick={userLogin}>
                  Login
                </Button>
                <Link to="/register">
                  <Text textAlign={"center"}>Create an account</Text>
                </Link>
              </Box>
            </Stack>
          </CardBody>
        </Card>
      </Container>
    </Box>
  );
};
