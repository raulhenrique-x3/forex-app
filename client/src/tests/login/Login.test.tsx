import { Route, Routes } from "react-router";
import { BrowserRouter } from "react-router-dom";
import { AuthContextProvider } from "../../context/AuthContext";
import { Login } from "../../pages/login/Login";
import { render, cleanup, act } from "@testing-library/react";
import { screen } from "@testing-library/dom";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";

jest.setTimeout(60000);
afterEach(cleanup);

describe("User access the login page", () => {
  it("should have inputs of email & password & a button to login", async () => {
    await act(async () => {
      render(
        <AuthContextProvider>
          <BrowserRouter>
            <Routes>
              <Route path="/" element={<Login submit={() => {}} />} />
            </Routes>
          </BrowserRouter>
        </AuthContextProvider>
      );
    });
    const email = screen.queryByPlaceholderText("Enter your email");
    const password = screen.queryByPlaceholderText("Enter your password");
    const button = screen.getByTestId("loginButton");

    expect(email).toBeInTheDocument();
    expect(password).toBeInTheDocument();
    expect(button).toBeInTheDocument();
  });

  it("should allow the user to submit heir credentials", () => {
    const login = jest.fn();
    render(
      <AuthContextProvider>
        <BrowserRouter>
          <Routes>
            <Route path="/" element={<Login submit={login} />} />
          </Routes>
        </BrowserRouter>
      </AuthContextProvider>
    );
    const email = screen.getByPlaceholderText("Enter your email");
    const password = screen.getByPlaceholderText("Enter your password");
    const button = screen.getByTestId("loginButton");

    userEvent.type(email, "raul@email.com");
    userEvent.type(password, "123456789");
    userEvent.click(button);

    expect(login).toHaveBeenCalledWith({ userEmail: "raul@email.com", userPassword: "123456789" });
  });
});
