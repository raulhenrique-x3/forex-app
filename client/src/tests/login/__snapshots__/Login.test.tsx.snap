// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
[
  <div
    className="css-l7yoyi"
  >
    <div
      className="chakra-container css-vhc7ah"
    >
      <div
        className="chakra-card css-jv498s"
      >
        <div
          className="chakra-card__header css-1sl53ol"
        >
          <p
            className="chakra-text css-1ug71u4"
          >
            Login
          </p>
        </div>
        <div
          className="chakra-card__body css-1idwstw"
        >
          <div
            className="chakra-stack css-j7qwjs"
          >
            <div
              className="css-0"
            >
              <label
                className="chakra-form__label css-g6pte"
              >
                Email
              </label>
              <input
                className="chakra-input css-1kp110w"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Enter your email"
                type="text"
              />
            </div>
            <div
              className="chakra-stack__divider css-1q5cbl"
            />
            <div
              className="css-0"
            >
              <p
                className="chakra-text css-0"
              >
                Password
              </p>
              <input
                className="chakra-input css-1kp110w"
                id="password"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Enter your password"
                type="password"
              />
            </div>
            <div
              className="chakra-stack__divider css-1q5cbl"
            />
            <div
              className="css-j7qwjs"
            >
              <button
                className="chakra-button css-1uxbacm"
                id="loginButton"
                onClick={[Function]}
                type="button"
              >
                Login
              </button>
              <a
                href="/register"
                onClick={[Function]}
              >
                <div
                  className="chakra-container css-hpbio4"
                >
                  <p
                    className="chakra-text css-kf7p74"
                  >
                    Create an account
                  </p>
                </div>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <span />,
  <span
    hidden={true}
    id="__chakra_env"
  />,
]
`;
